fq.go

```diff
@@ -0,0 +1,163 @@
+package fq
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+)
+
+type FQScheduler struct {
+	lock   *sync.Mutex
+	queues []*Queue
+	vt     uint64
+	C      uint64
+	G      uint64
+	seen   bool
+}
+
+// TODO(aaron-prindle) add concurrency enforcement - 'C'
+
+func (q *FQScheduler) chooseQueue(packet *Packet) *Queue {
+	for _, queue := range q.queues {
+		if packet.key == queue.key {
+			return queue
+		}
+	}
+	panic("no matching queue for packet")
+}
+
+func newFQScheduler(queues []*Queue) *FQScheduler {
+	now := NowAsUnixMilli()
+	fq := &FQScheduler{
+		lock:   &sync.Mutex{},
+		queues: queues,
+		// R(t) = (server start time) + (1 ns) * (number of rounds since server start).
+		vt: now,
+	}
+	// TODO(aaron-prindle) verify if this is needed?
+	for i := range fq.queues {
+		fq.queues[i].virstart = fq.vt
+	}
+
+	return fq
+}
+
+// TODO(aaron-prindle) verify that the time units are correct/matching
+func NowAsUnixMilli() uint64 {
+	return uint64(time.Now().UnixNano() / 1e6)
+}
+
+func (q *FQScheduler) processround() (*Packet, bool) {
```

Comment:
Is this function intended for anything besides testing?  It is not clear to me how this function would be used in the real context, because this function bundles together two unrelated things: advancing the virtual clock by a fixed "tick", and dequeuing one request.

Reply:
My understanding was that a round of fair queuing advanced the virtual time and dequeud one request.  "processround"  was meant to be advance a round.  

> this function bundles together two unrelated things: advancing the virtual clock by a fixed "tick", and dequeuing one request.

I thought these were related.  Can you explain when a "tick" occurs or more about how a round is structured?  I thought that in one "round" we identify how much work/time we are advancing (dvt/dt) and then we dequeue one request

Reply:
In Demers et al paper, a round consists of transmitting one bit from each non-empty queue.  A round is purely an imaginary concept.  The packets are not transmitted bit-by-bit, and there is no code that runs once per round.  In fact, no code is even scheduled based on any calculation with virtual time AKA rounds.  There is code that runs when a packet arrives, and there is code that runs when the packet currently being transmitted finishes.  It is the arrivals and completions of packets that set the real timing.

In our adaptation, a round consists of giving one nanosecond (or whatever increment we prefer to imagine) of CPU to every non-empty queue.  Again, nothing runs once per round.  We need code that runs when a request arrives, and code that runs when a request is completed.  We need a function that advances the virtual clock by the right amount of virtual time since it was last adjusted, based on the amount of real time that passed and that partial derivative (which will be constant since the last adjustment, since the derivative only changes when our code runs(.